spring-integration-core/src/main/java/org/springframework/integration/transformer/FromProtobufTransformer.java

```diff
@@ -0,0 +1,160 @@
+	 */
+	public FromProtobufTransformer(Class<? extends com.google.protobuf.Message> expectedType,
+			ProtobufMessageConverter protobufMessageConverter) {
+		Assert.notNull(expectedType, "'defaultType' must not be null");
+		Assert.notNull(protobufMessageConverter, "'protobufMessageConverter' must not be null");
+		this.expectedType = expectedType;
+		this.protobufMessageConverter = protobufMessageConverter;
+	}
+
+	@Override
+	public void setBeanClassLoader(ClassLoader classLoader) {
+		this.beanClassLoader = classLoader;
+	}
+
+	/**
+	 * Set the expected protobuf class type.
+	 * @param expectedType expected protobuf class type.
+	 */
+	public void setExpectedType(Class<? extends com.google.protobuf.Message> expectedType) {
+		this.expectedType = expectedType;
+	}
+
+	/**
+	 * Get the Protobuf expected class, if configured.
+	 * @return returns the expected Protocol Buffer class.
+	 */
+	public Class<? extends com.google.protobuf.Message> getExpectedType() {
+		return expectedType;
+	}
+
+	/**
+	 * Set the expression to evaluate against the message to determine the type. Default {@code headers['proto_type']}.
+	 * @param expression the expression.
+	 */
+	public void setExpectedTypeExpression(Expression expression) {
+		Assert.notNull(expression, "'expression' must not be null");
+		this.expectedTypeExpression = expression;
+	}
+
+	/**
+	 * Set the expression to evaluate against the message to determine the type id. Default
+	 * {@code headers['proto_type']}.
+	 * @param expression the expression.
+	 */
+	public void setExpectedTypeExpressionString(String expression) {
+		Assert.notNull(expression, "'expression' must not be null");
+		this.expectedTypeExpression = EXPRESSION_PARSER.parseExpression(expression);
+	}
+
+	@Override
+	protected void onInit() {
+		this.evaluationContext = IntegrationContextUtils.getEvaluationContext(getBeanFactory());
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	protected Object doTransform(Message<?> message) {
+
+		Class<? extends com.google.protobuf.Message> targetClass = null;
+		Object value = this.expectedTypeExpression.getValue(this.evaluationContext, message);
+		if (value instanceof Class) {
+			targetClass = (Class<? extends com.google.protobuf.Message>) value;
+		}
+		else if (value instanceof String) {
+			try {
+				targetClass = (Class<? extends com.google.protobuf.Message>) ClassUtils.forName((String) value,
+						this.beanClassLoader);
+			}
+			catch (ClassNotFoundException | LinkageError e) {
+				throw new IllegalStateException(e);
+			}
+		}
+		if (targetClass == null) {
+			targetClass = this.expectedType;
+		}
+
+		if (targetClass == null) {
+			this.logger.error(() -> "The 'expectedTypeExpression' (" + this.expectedTypeExpression
+					+ ") returned 'null' for: "
+					+ message + ". No falling back expectedType is configured. Consider setting the expectedType.");
+			throw new RuntimeException("The 'expectedTypeExpression' (" + this.expectedTypeExpression
```

Comment:
The exception must be `MessageTransformationException` to shorten stack trace a little bit upstream.